networkit/cpp/generators/test/GeneratorsGTest.cpp

```diff
@@ -172,15 +173,16 @@ TEST_F(GeneratorsGTest, testStaticPubWebGenerator) {
 	DEBUG("Modularity of clustering: " , modVal);
 	DEBUG("Total edge weight: " , G.totalEdgeWeight());
 	EXPECT_TRUE(G.checkConsistency());
+
+	//reset seed to previous
+	Aux::Random::setSeed(seed, false);
```

Comment:
This does not restore the random number generator to the previous state but instead restores it to the previously set static seed. I don't understand why this is done here. Could you please explain why the seed should be restored? If there had been no seed set previously, this will set a seed but the generator will be in a completely different state. If there had been a seed set, it restores the random number generator to this state. However, I don't see why anyone would want to have their seed restored.

Imho every test that uses randomness should set a seed in order to ensure that test results are reproducible. Otherwise test results are random/depend on how many and which other tests were executed before the current test.

Reply:
Well, as far as I understand the concept of this test suite, there is an initially non static seed at the beginning, but because some tests require (or at least is preferred to) set a see, this non static seed is overwritten with some other value e.g. 42. 

The problem here is that a following test, which is written in mind that there is a non static seed, is now run with a seed of 42 because the previous test set it so. 

Sooner or later, the whole test suite runs with a static seed because some test initially set it to a specific value. Ideally I would like to have every test to pass with a non-static seed (I see that this is not possible for every test), but this behaviour is impossible at the moment.

I admit, that this function is not part of the branch at the moment, but I think a good design would be to run the whole test suite with a random, but static seed which is printed in front of the output. This way, test which are well written run with a truly random seed while others can set their own seed. In addition, if a test fails for a random seed one can reproduce this bug because the seed is provided in the output.

Ideally this behaviour can also be implemented without this construct of preserving the seed (e.g by implemented an automatic reset after every test).

Reply:
While I agree that tests should be pass with random seeds, imho tests should also be reproducible. It is not nice if you implement a feature, submit a pull request and then some random, completely unrelated unit test fails because of the random seed that was used.

Having the possibility to specify a static seed and otherwise printing the used random seed is indeed a good alternative as it means random test failures are reproducible. Note that this would need to happen before every single test function as otherwise the order and number of tests that are executed plays a role (and one shouldn't need to re-run the whole test suite to reproduce a failure). Further, there might be algorithms that uses randomness inside multiple threads that use a different number of random numbers/thread depending on the scheduling which might leave the random generator in a state that depends on the scheduling afterwards. For continuous integration, one could also just use a fixed seed in order to avoid the unrelated test failures described above.

Some time ago, fixed seeds were added in NetworKit at least to some tests in order to make the test suite more reproducible and to avoid random test failures. You are right that they have effects on other tests, too. However, your change does not avoid these effects. It just sets the random number generator to a different seed. Either to the previously set seed, or, if there was no previous seed, to a random seed. Note, however, that you only get a random seed once. After setting a seed once, `getSeed` will always return the previously set seed.

My personal suggestion would be to leave static seeds as they are currently (i.e., remove the changes to random seeds from this PR) and implement random seeds that are reset for every test and printed later.

Reply:
I agree that tests should be reproducible. On the other hand, they should not hard code results that depend on a random seed. A test that checks if some algorithm output equals X if random seed S was used is useless (as X was clearly obtained by just running the algorithm with the given seed) and will break if the random number generator changes.

We should remove all cases of the latter behavior from the test suite. Still, we should use random seeds for every test to avoid the spurious failures that @michitux mentioned. If it is feasible, we can loop in a test and check different fixed random seeds.

Reply:
https://github.com/kit-parco/networkit/pull/170/commits/75c0e0fd4cc6761505f8e0027ee8be0ed1469315 reverts the changes